sdks/rust/src/sdk.rs

```diff
@@ -12,349 +12,211 @@
+    /// Starts a new SDK instance, and connects to localhost on the port specified
+    /// or else falls back to the `AGONES_SDK_GRPC_PORT` environment variable,
+    /// or defaults to 9357.
+    ///
+    /// Note that this function will loop indefinitely until a connection is made
+    /// to the SDK server, so it is recommended to wrap this in a
+    /// [timeout](https://docs.rs/tokio/1.6.0/tokio/time/fn.timeout.html).
+    pub async fn new(port: Option<u16>, keep_alive: Option<Duration>) -> Result<Self> {
+        // TODO: Add TLS? For some reason the original Cargo.toml was enabling
+        // grpcio's openssl features, but AFAICT the SDK and sidecar only ever
+        // communicate via a non-TLS connection, so seems like we could just
+        // use the simpler client setup code if TLS is absolutely never needed
+        let addr: http::Uri = format!(
+            "http://localhost:{}",
+            port.unwrap_or_else(|| {
+                env::var("AGONES_SDK_GRPC_PORT")
+                    .ok()
+                    .and_then(|s| s.parse().ok())
+                    .unwrap_or(9357)
+            })
+        )
+        .parse()?;
+
+        let builder = tonic::transport::channel::Channel::builder(addr)
+            .keep_alive_timeout(keep_alive.unwrap_or_else(|| Duration::from_secs(30)));
+
+        // let mut root_store = rustls::RootCertStore::empty();
+        // root_store.add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
+        // let mut rusttls_config = rustls::ClientConfig::new();
+        // rusttls_config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+        // rusttls_config.root_store = root_store;
+        // let tls_config =
+        //     tonic::transport::ClientTlsConfig::new().rustls_client_config(rusttls_config);
+        // builder = builder.tls_config(tls_config)?;
+
+        let channel = builder.connect().await?;
+        let mut client = SdkClient::new(channel.clone());
+        let alpha = Alpha::new(channel);
+
+        // Loop until we connect. The original implementation just looped once
+        // every second up to a maximum of 30 seconds, but it's better for the
+        // external caller to wrap this in their own timeout
+        let mut connect_interval = tokio::time::interval(Duration::from_millis(100));
 
-        // Unfortunately there isn't a native way to block until connected
-        // so we had to roll our own.
-        let mut counter = 0;
         loop {
-            counter += 1;
-            match cli.get_game_server(&req) {
-                Ok(_) => break,
-                Err(e) => {
-                    if counter > 30 {
-                        return Err(ErrorKind::Grpc(e).into());
-                    }
-                    sleep(Duration::from_secs(1));
-                    continue;
-                }
+            connect_interval.tick().await;
+
+            if client.get_game_server(empty()).await.is_ok() {
+                break;
```

Comment:
Like the comment says, this changes this function to loop indefinitely instead of using a hardcoded timeout. It's generally better in async code to let the caller decide on the timeout, which in this case can be done with [tokio::time::timeout](https://docs.rs/tokio/1.6.0/tokio/time/fn.timeout.html).

Reply:
All our other SDK clients fail after 30s, I think we should probably be consistent here.